firebase-storage/src/main/java/com/google/firebase/storage/ListResult.java

```diff
@@ -0,0 +1,118 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.storage;
+
+import android.support.annotation.Nullable;
+import com.google.android.gms.common.internal.Preconditions;
+import com.google.firebase.annotations.PublicApi;
+import java.util.ArrayList;
+import java.util.List;
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+/** A single paged result of a {@link StorageReference#list} call. */
+@PublicApi
+public class ListResult {
+  private static final String ITEMS_KEY = "items";
+  private static final String NAME_KEY = "name";
+  private static final String PAGE_TOKEN_KEY = "nextPageToken";
+  private static final String PREFIXES_KEY = "prefixes";
+
+  private List<StorageReference> prefixes;
+  private List<StorageReference> items;
+  @Nullable private String pageToken;
+
+  ListResult(
+      List<StorageReference> prefixes, List<StorageReference> items, @Nullable String pageToken) {
+    this.prefixes = prefixes;
+    this.items = items;
+    this.pageToken = pageToken;
+  }
+
+  /**
+   * The prefixes (folders) returned by the {@code list()} operation.
+   *
+   * @return A list of prefixes (folders).
+   */
+  @PublicApi
+  public List<StorageReference> getPrefixes() {
+    return prefixes;
+  }
+
+  /**
+   * The items (files) returned by the {@code list()} operation.
+   *
+   * @return A list of items (files).
+   */
+  @PublicApi
+  public List<StorageReference> getItems() {
+    return items;
+  }
+
+  /**
+   * Returns a token that can be used to resume a previous {@code list()} operation. `null`
+   * indicates that there are no more results.
+   *
+   * @return A page token if more results are avaible.
+   */
+  @PublicApi
+  @Nullable
+  public String getPageToken() {
+    return pageToken;
+  }
+
+  static class Builder {
+    private final JSONObject resultBody;
+    private final StorageReference storageRef;
+    private final FirebaseStorage firebaseStorage;
+
+    public Builder(JSONObject resultBody, StorageReference storageRef) {
```

Comment:
I would suggest pass-in `StorageReference` in the constructor, and have `appendResponse(JSONObject )` to parse and append results. This way, `listAll` can take advantage of the `Builder`.
The fact that `build()` might blow up is quite surprising.

nit: Google3 Java usually add a static `toBuilder(StorageReference storageRef)`. No strong opinion here.